src/utils/serdes/serdes.cpp

```diff
@@ -14,121 +14,156 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#include <stdexcept>
+#include <utility>
+
 #include "serdes.h"
 
-nixlSerDes::nixlSerDes() {
-    workingStr = "nixlSerDes|";
-    des_offset = 11;
+namespace
+{
+   const std::string_view introString = "nixlSerDes|";
+
+}  // namespace
+
+nixlSerializer::nixlSerializer()
+    : buffer_(introString)
+{}
 
-    mode = SERIALIZE;
+nixlSerializer::nixlSerializer(const std::size_t pre)
+{
+    buffer_.reserve(pre);
+    buffer_.append(introString);
 }
 
-std::string nixlSerDes::_bytesToString(const void *buf, ssize_t size) {
-    std::string ret_str = std::string(reinterpret_cast<const char*>(buf), size);
-    return ret_str;
+void nixlSerializer::addRaw(const std::string_view& str)
+{
+    addRaw(str.data(), str.size());
 }
 
-void nixlSerDes::_stringToBytes(void* fill_buf, const std::string &s, ssize_t size){
-    s.copy(reinterpret_cast<char*>(fill_buf), size);
+void nixlSerializer::addRaw(const void* buf, const std::size_t len)
+{
+    buffer_.append(reinterpret_cast<const char*>(buf), len);
 }
 
-// Strings serialization
-nixl_status_t nixlSerDes::addStr(const std::string &tag, const std::string &str){
+void nixlSerializer::addStr(const std::string_view& tag, const std::string_view& str)
+{
+    addBuf(tag, str.data(), str.size());
+}
 
-    size_t len = str.size();
+void nixlSerializer::addBuf(const std::string_view& tag, const void* buf, std::size_t len)
+{
+    buffer_.reserve(buffer_.size() + tag.size() + sizeof(len) + len + 1 );
 
-    workingStr.append(tag);
-    workingStr.append(_bytesToString(&len, sizeof(size_t)));
-    workingStr.append(str);
-    workingStr.append("|");
+    addRaw(tag);
+    addRaw(&len, sizeof(len));
+    addRaw(buf, len);
+    addRaw("|");
+}
 
-    return NIXL_SUCCESS;
+std::string nixlSerializer::exportStr() && noexcept
+{
+    return std::move(buffer_);
 }
 
-std::string nixlSerDes::getStr(const std::string &tag){
+const std::string& nixlSerializer::exportStr() const & noexcept
+{
+    return buffer_;
+}
 
-    if(workingStr.compare(des_offset, tag.size(), tag) != 0){
-       //incorrect tag
-       return "";
+nixlDeserializer::nixlDeserializer(std::string&& str)
+    : buffer_(std::move(str)),
+      offset_(introString.size())
+{
+    if(buffer_.size() < introString.size() ) {
+        throw std::runtime_error("insufficient deserialization data for intro");
+    }
+    if(std::memcmp(buffer_.data(), introString.data(), introString.size() != 0 )) {
```

Comment:
It looks like '!= 0' here is on the wrong side of the first closing bracket, i.e. it compares string size to 0, not the result of memcmp function.